File: line_2d3d/python/MyTriangle.py
from __future__ import annotations
from typing import List
from MyUtils import *
from MyCamera import *
from MyColor import *
from MyLine import *
import numpy as np
from MyUtils import *
from MyLight import *

class MyTriangle :
#   MyLine[] ls;
#   PVector p1, p2, p3, n, up;  // plane : nx*x + ny*y + nz*z = o
#   float   o;
#   color   c;
#   float   shading;
#   int     id;



  def __init__(self, p1:List[float], p2:List[float], p3:List[float], vis1:bool, vis2:bool, vis3:bool, up:List[float], light_n:MyLight, lines:List[MyLine], cam:MyCamera ) :

    self.id = getID()
    self.p1 = cam.project(p1)
    self.p2 = cam.project(p2)
    self.p3 = cam.project(p3)
    self.up = up

    self.det_normal_and_o()

    self.ls = [MyLine]*3
    self.ls[0] = MyLine(self, self.p1, self.p2, MyColor(255), vis1)
    self.ls[1] = MyLine(self, self.p2, self.p3, MyColor(255), vis2)
    self.ls[2] = MyLine(self, self.p3, self.p1, MyColor(255), vis3)

    for i in range(3) :
      lines.append(self.ls[i])

    self.c = MyColor()

    self.set_light(light_n)
  

  def __str__(self) :
    return f"t: if : {self.id} n  {self.n} o: {self.o}"

  def draw(self, svg) :
    for i in range(3) :
      self.ls[i].draw(svg)

  def draw_normal(self, svg) :
    c = self.center()
    cn = np.add(c,np.multiply(n, 50))
    svg.line(str(MyColor(255,0,0)), self.thickness, c[X], c[Y], cn[X], cn[Y] )

  def equals(self, t:MyTriangle) -> bool:
    return self.id == t.id


  def triangleAreaXY(self, p1:List[float], p2:List[float], p3:List[float]) :
    a = abs(p1[X]*(p2[Y]-p3[Y]) + p2[X]*(p3[Y]-p1[Y]) + p3[X]*(p1[Y]-p2[Y]))/2.0
    return a
  

#   //if include_edge is false : if the point is on the edge, is NOT inside. 
#   //if include_edge is true  : if the point is on the edge, is inside. 
  def insideTriangleXY(self, p:List[float], include_edge : bool) -> bool :
    area  = self.triangleAreaXY(self.p1, self.p2, self.p3)
    area1 = self.triangleAreaXY(p, self.p2, self.p3)
    area2 = self.triangleAreaXY(self.p1, p, self.p3)
    area3 = self.triangleAreaXY(self.p1, self.p2, p)
    if not include_edge :
      if (area1 <= FLOATING_POINT_ACCURACY) or (area2 <= FLOATING_POINT_ACCURACY) or (area3 <= FLOATING_POINT_ACCURACY) :
        return False
    ins = (abs(area1+area2+area3) - area) <= (FLOATING_POINT_ACCURACY * area)
    return ins
  

  def det_normal_and_o(self) :
    s = np.subtract(self.p2, self.p1)
    t = np.subtract(self.p3, self.p2)
    n = np.cross(s,t )
    self.n = n / np.linalg.norm(n)
    self.o = np.dot(self.n, self.p1)
  

  def center(self) -> List[float] :
    p = np.add(self.p1, self.p2)
    p = np.add(p, self.p3)
    p = np.multiply(p, (1.0/3.0))
    return p
  

#   /**
#    * #TODO DIT IS NOG NIET GOED>
#    */
#   MyLine planePlaneIntersect(MyTriangle pl2) {
#     //https://mathemerize.com/equation-of-plane-passing-through-intersection-of-two-planes/
#     //https://www[Y]outube.com/watch?v=O6O_64zIEYI

#     PVector v = n.cross( pl2.n);

#     float o1 = o;
#     float o2 = pl2.o;
#     PVector n1 = n;
#     PVector n2 = pl2.n;

#     if ( PVector.sub(n1, n2).dot(1, 1, 1) < FLOATING_POINT_ACCURACY) {
#       println("Parallel");
#       return null;
#     }

#     MyLine l;
#     if ((abs(n1[X]) < FLOATING_POINT_ACCURACY) && (abs(n1[Y]) < FLOATING_POINT_ACCURACY)) {
#       // we cannot take z = 0; so we take y = 0
#       println("plane 1 parallel z plane");
#       float teller = o1 - o2 + n2.dot(v) - n1.dot(v);
#       float noemer = n1[X] - (n2[X] * n1[Z]) / n2[Z];
#       float px = teller / noemer;
#       float pz = (o2 - n2.dot(v) - n2[X]*px ) / n2[Z];

#       PVector p = new PVector(px, 0, pz);
#       l = new MyLine(this, v, p, color(255, 0, 0), 5, true);
#       println("found line : " +l);
#     } else {
#       // z = 0
#       println("plane 1 crosses z plane");

#       float teller = o1 - o2 + n2.dot(v) - n1.dot(v);
#       float noemer = n1[X] - (n2[X] * n1[Y]) / n2[Y];
#       float px = teller / noemer;
#       float py = (o2 - n2.dot(v) - n2[X]*px ) / n2[Y];

#       PVector p = new PVector(px, py, 0);
#       l = new MyLine(this, v, p, color(255, 0, 0), 5, true);
#       println("found line : " +l);
#     }

#     return l;
#   }

  def getZ(self, x:float, y:float) -> float :
    if (self.n[Z] != 0):
       return (self.o - self.n[X]*x - self.n[Y]*y)/self.n[Z]
    else :
        return 0

  def set_light(self, l : MyLight) :
    self.shading = np.dot(self.n, l.getNormalizedDirection())
    if (self.shading > 0) :
      self.shading = 0
    else:
      self.shading = -self.shading
    

  def getHatches(self) -> List[MyLine]:
    hatches = []
    hatch_min = 5
    hatch_grad = 15

    if np.dot(self.n, (0,0,1)) <= 0 :
    #   // no need to hatch the backside of a vertex
      return hatches
    


    left = self.p1
    if (self.p2[X] < left[X]) : left = self.p2
    if (self.p3[X] < left[X]) : left = self.p3
    right = self.p1
    if (self.p2[X] > right[X]) : right = self.p2
    if (self.p3[X] > right[X]) : right = self.p3

    bottom = self.p1
    if (self.p2[Y] < bottom[Y]) : bottom = self.p2
    if (self.p3[Y] < bottom[Y]) : bottom = self.p3
    top = self.p1
    if (self.p2[Y] > top[Y]) : top = self.p2
    if (self.p3[Y] > top[Y]) : top = self.p3

    # // make a square around the triangle
    sqr_bl = [0]*2
    sqr_bl[X] = left[X]
    sqr_bl[Y] = bottom[Y]
    sqr_tr = [0]*2
    sqr_tr[X] = right[X]
    sqr_tr[Y] = top[Y]

    # // https://www.tutorialspoint.com/Check-whether-a-given-point-lies-inside-a-Triangle

    



    hatch_spacing =  hatch_min + self.shading*hatch_grad; #pixels, to be replaces with shading
    x = sqr_bl[X]
    while x < sqr_tr[X]:
      b = [x, bottom[Y] - FLOATING_POINT_ACCURACY, 0]
      b[Z] = self.getZ(b[X], b[Y])
      t = [x, top[Y] + FLOATING_POINT_ACCURACY, 0]
      t[Z] = self.getZ(t[X], t[Y])

      hatch_line = MyLine(self, b, t)

      intersects = []  #// intersectes within the square/triangle
      for i in range(3):
        is_is, i_s = hatch_line.getLineIntersectionXY(self.ls[i])
        if is_is:
          is_in = self.insideTriangleXY(i_s, True)
          if is_in:
            intersects.extend(i_s)
      
      if len(intersects) == 3 :
        # // remove duplicate: probably  a triangle corner
        hash_set = set(intersects)
        intersects = list(hash_set)
      
      
      if len(intersects) == 2 :
        hatches.extend(MyLine(self, intersects[0], intersects[1], MyColor(255)))
      

      x += hatch_spacing

    return hatches

File: line_2d3d/cpp/line_2d3d/line_2d3d/MyTriangle.cpp
#include <format>
#include <iostream>
#include "MyTriangle.h"
#include "MyUtil.h"


void MyTriangle::create(Vector3d p1_arg, Vector3d p2_arg, Vector3d p3_arg, bool vis1, bool vis2, bool vis3, Vector3d up_arg, MyLight& light_n, MyCamera& cam)
{
    id = getID();
    ps[0] = cam.project(p1_arg);
    ps[1] = cam.project(p2_arg);
    ps[2] = cam.project(p3_arg);
    up = cam.project(up_arg);

    // plane vectors
    det_normal_and_o();




    ls[0] = MyLine(id, ps[0], ps[1], MyColor(0), vis1);
    ls[1] = MyLine(id, ps[1], ps[2], MyColor(0), vis2);
    ls[2] = MyLine(id, ps[2], ps[0], MyColor(0), vis3);

    for (int i = 0; i < 3; i++) {
        lines.push_back(ls[i]);
    }

    ///set_light(light_n);
    shading = n.dot(light_n.getNormalizedDirection());
    if (shading > 0) {
        shading = 0;
    }
    else {
        shading = -shading;
    }

    addHatching3D(up);

}

void MyTriangle::det_normal_and_o()
{
    Vector3d s = ps[1] - ps[0];
    Vector3d t = ps[2] - ps[1];
    n = s.cross(t);
    n.normalize();
    o = n.dot(ps[0]);
    right = n.cross(up);
    //cout << "det_normal_and_o: n " << n.transpose() << " o " << o << "\n";
}


MyTriangle::MyTriangle(Vector3d p1_arg, Vector3d p2_arg, Vector3d p3_arg, bool vis1, bool vis2, bool vis3, Vector3d up_arg, MyLight& light_n, MyCamera& cam)
{
    create(p1_arg, p2_arg, p3_arg, vis1, vis2, vis3, up_arg, light_n, cam);
}

MyTriangle::MyTriangle(Vector3d p1, Vector3d p2, Vector3d p3, Vector3d up, MyLight& light_n, MyCamera& cam)
{
    create(p1, p2, p3, true, true, true, up, light_n, cam);
}

void MyTriangle::draw(MySvg &svg)
{

    for (MyLine& line : lines) {
        line.draw(svg);
    }

    // draw center normal, up en right
    if (false) {
        Vector3d c = center();
        MyLine ml_n = MyLine(id, c, c + 100 * n, MyColor(255, 0, 0), 1, true);
        ml_n.draw(svg);
        MyLine ml_up = MyLine(id, c, c + 100 * up, MyColor(0, 255, 0), 1, true);
        ml_up.draw(svg);
        MyLine ml_right = MyLine(id, c, c + 100 * right, MyColor(0, 0, 255), 1, true);
        ml_right.draw(svg);
    }
}

void MyTriangle::draw_normal(MySvg svg)
{
    Vector3d c = center();
    Vector3d cn = n * 50;
    svg.line(MyColor(255,0,0).str(), 1, c[0], c[1], cn[0], cn[1]);
}

string MyTriangle::str()
{
    //return format("id {} n {} o {}", id, n, o);
    return string("roland");
}



bool MyTriangle::equals(MyTriangle& t)
{
    return id == t.id;
}

Vector3d MyTriangle::center()
{
    Vector3d c = (ps[0] + ps[1] + ps[2]) / 3.0;
    return c;
}

double MyTriangle::triangleAreaXY(Vector3d p1, Vector3d p2, Vector3d p3)
{
    return abs((p1[0] * (p2[1] - p3[1]) + p2[0] * (p3[1] - p1[1]) + p3[0] * (p1[1] - p2[1])) / 2.0);
}

//if include_edge is false : if the point is on the edge, is NOT inside. 
//if include_edge is true  : if the point is on the edge, is inside. 
bool MyTriangle::insideTriangleXY(Vector3d po, bool include_edge)
{
    double area = triangleAreaXY(ps[0], ps[1], ps[2]);
    double area1 = triangleAreaXY(po, ps[1], ps[2]);
    double area2 = triangleAreaXY(ps[0], po, ps[2]);
    double area3 = triangleAreaXY(ps[0], ps[1], po);
    if (!include_edge) {
        if ((area1 <= FLOATING_POINT_ACCURACY) || (area2 <= FLOATING_POINT_ACCURACY) || (area3 <= FLOATING_POINT_ACCURACY))
        {
            return false;
        }
    }
    return ((abs(area1 + area2 + area3) - area) <= FLOATING_POINT_ACCURACY * area);
}

double MyTriangle::getZ(double x, double y)
{
    if (n[2] != 0) {
        return (o - n[0] * x - n[1] * y) / n[2];
    }
    return 0;

}


/***
* add for all lines in this triangle the intersections in 3d with argument triangle
*/
int MyTriangle::addTriangleIntersectXYZ(MyTriangle& triangle)
{
    int no_intersects = 0;
    if (triangle.id != id) {
        for (MyLine &line : lines) {
            // https://en.wikipedia.org/wiki/Line%E2%80%93plane_intersection
            Vector3d p0 = triangle.ls[0].ps[0];
            Vector3d l0 = line.ps[0];
            Vector3d l = line.get_direction();
            double teller = (p0 - l0).dot(triangle.n);
            double noemer = l.dot(triangle.n);
            if (abs(noemer) > FLOATING_POINT_ACCURACY) {
                double d = teller / noemer;
                Vector3d p = l0 + l * d;
                if (triangle.insideTriangleXY(p, true) && (d < 1.0) && (d > 0.0)) {
                    line.addSplitter(d);
                    no_intersects++;
                }
            }
        }
    }

    return no_intersects;

}

/**
* For each line in this triangle add obscuratoin by other triangle.
*/
void MyTriangle::addTriangleObscuration(MyTriangle& triangle)
{
    if (triangle.id != id) {
        for (MyLine &line : lines) {
            addTriangleObscuration(triangle, line);
            for (MyLine &split_line : line.split_lines) {
                addTriangleObscuration(triangle, split_line);
            }
        }
    }
}


void MyTriangle::addTriangleObscuration(MyTriangle& triangle, MyLine& line) {
    Vector3d p = line.ps[0] + line.get_direction() * 0.5;
    double tz = triangle.getZ(p[0], p[1]);

    if (triangle.insideTriangleXY(p, false) && floatSmallerThenRelative(p[2], tz,  FLOATING_POINT_ACCURACY)) {
        line.visible = false;
    }
}


int MyTriangle::getNoVisLines()
{
    int no_vis_lines = 0;
    for (MyLine &line : lines) {
        no_vis_lines += line.getNoVisibleLines();
    }
    return no_vis_lines;
}

int MyTriangle::getVisLines(vector<MyLine>& vis_lines) {
    int no_vis_lines = 0;
    for (MyLine& line : lines) {
        no_vis_lines += line.getVisibleLines(vis_lines);
    }
    return no_vis_lines;
}

/**
* add line intersects to the lines of this triangle wrt to the triangle argument main lines
*/
int MyTriangle::addLineLineIntersectionXY(MyTriangle& triangle)
{
    int no_intersects = 0;
    if (triangle.id != id) {
        for (MyLine& line : lines) {
            for (MyLine &main_line : triangle.ls) {
                Vector3d intersect;
                if (line.addLineIntersectionXY(main_line, intersect)) {
                    no_intersects++;
                }
            }
        }
    }

    return no_intersects;

}

void MyTriangle::generateSplitLines()
{
    for (MyLine& line : lines) {
        line.generateSplitLines();
    }
}

void MyTriangle::recombineLines()
{
    for (MyLine& line : lines) {
        line.recombineLines();
    }
}

void MyTriangle::addHatching3D(Vector3d up) {


    //no backside shading
    if (n.dot(Vector3d(0, 0, 1)) <= 0) {
        return;
    }

    // lowest and highest vector wrt up.
    Vector3d up_pos = Vector3d(0,0,0);
    for (int i = 0; i < 3; i++) {
        up_pos[i] = up.dot(ps[i]);
    }

    int smallest_i = 0;
    int biggest_i = 0;
    for (int i = 1; i < 3; i++) {
        if (up_pos[i] < up_pos[smallest_i]) {
            smallest_i = i;
        } 
        if (up_pos[i] > up_pos[smallest_i]) {
            biggest_i = i;
        }
    }
    // start position
    double hatch_min = 5;
    double hatch_grad = 25;
    double hatch_spacing = hatch_min + shading * hatch_grad; //pixels, to be replaced with shading
    double hatch_start = up_pos[smallest_i] -fmod(up_pos[smallest_i], hatch_spacing);
    int i = 0;
    for (double x = 0; i < 1000; x += hatch_spacing) {
        i++;
        Vector3d b = ps[smallest_i] + x * up;
        Vector3d t = ps[smallest_i] + x * up - 100*right;
        MyLine hatch_line = MyLine(id, b, t);

        vector<Vector3d> intersections;  // intersectes within the square/triangle
        for (int i = 0; i < 3; i++) {
            Vector3d intersect;
            if (hatch_line.getLineIntersectionXY(ls[i], intersect, true)) {
                bool is_in = insideTriangleXY(intersect, true);
                if (is_in) {
                    intersections.push_back(intersect);
                }
            }
        }

        if (intersections.size() == 0) {
            break;
        }

        if (intersections.size() == 3) {
            Vector3d delta = (intersections[0] - intersections[1]);
            if (floatEqualsRelative(0, delta.norm(), FLOATING_POINT_ACCURACY)) {
                intersections.erase(intersections.begin() + 1);
            }
            else {
                intersections.erase(intersections.begin() + 2);
            }
        }

        if (intersections.size() == 2) {
            MyLine line = MyLine(id, intersections[0], intersections[1], MyColor(0));
            lines.push_back(line);
        }

    }
}

void MyTriangle::addHatches()
{
    double hatch_min = 5;
    double hatch_grad = 25;
    double hatch_spacing = hatch_min + shading * hatch_grad; //pixels, to be replaced with shading


    if (n.dot(Vector3d(0, 0, 1)) <= 0) {
        return;
    }


    Vector3d left = ps[0];
    if (ps[1][0] < left[0]) left = ps[1];
    if (ps[2][0] < left[0]) left = ps[2];
    Vector3d right = ps[0];
    if (ps[1][0] > right[0]) right = ps[1];
    if (ps[2][0] > right[0]) right = ps[2];

    Vector3d bottom = ps[0];
    if (ps[1][1] < bottom[1]) bottom = ps[1];
    if (ps[2][1] < bottom[1]) bottom = ps[2];
    Vector3d top = ps[0];
    if (ps[1][1] > top[1]) top = ps[1];
    if (ps[2][1] > top[1]) top = ps[2];

    // make a square around the triangle
    Vector3d sqr_bl;
    sqr_bl[0] = left[0];
    sqr_bl[1] = bottom[1];
    Vector3d sqr_tr = Vector3d();
    sqr_tr[0] = right[0];
    sqr_tr[1] = top[1];



    //Vector3d view_dir = Vector3d.sub(cam_pos, look_at).normalize();
    //Vector3d side_dir = up.cross(view_dir).normalize();
    //Vector3d e = view_dir.cross(side_dir).normalize();

    
    // https://www.tutorialspoint.com/Check-whether-a-given-point-lies-inside-a-Triangle
    double hatch_start = sqr_bl[0] - fmod(sqr_bl[0], hatch_spacing);
    for (double x = hatch_start; x < sqr_tr[0]; x += hatch_spacing) {
        Vector3d b = Vector3d(x, bottom[1] - FLOATING_POINT_ACCURACY, 0);
        b[2] = getZ(b[0], b[1]);
        Vector3d t = Vector3d(x, top[1] + FLOATING_POINT_ACCURACY, 0);
        t[2] = getZ(t[0], t[1]);


        MyLine hatch_line = MyLine(id, b, t);

        vector<Vector3d> intersections;  // intersectes within the square/triangle
        for (int i = 0; i < 3; i++) {
            Vector3d intersect;
            if (hatch_line.getLineIntersectionXY(ls[i], intersect)) {
                bool is_in = insideTriangleXY(intersect, true);
                if (is_in) {
                    intersections.push_back(intersect);
                }
            }
        }

        if (intersections.size() == 3) {
            Vector3d delta = (intersections[0] - intersections[1]);
            if (floatEqualsRelative(0, delta.norm(), FLOATING_POINT_ACCURACY)) {
                intersections.erase(intersections.begin() + 1);
            } else {
                intersections.erase(intersections.begin() + 2);
            }
        }

        if (intersections.size() == 2) {
            lines.push_back(MyLine(id, intersections[0], intersections[1], MyColor(0)));
        }
    }

}

